dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/supplier/ProgramOwnerSupplier.java

```diff
@@ -0,0 +1,98 @@
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.tracker.preheat.supplier;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+
+import org.hisp.dhis.program.ProgramInstance;
+import org.hisp.dhis.trackedentity.TrackedEntityInstance;
+import org.hisp.dhis.trackedentity.TrackedEntityProgramOwnerOrgUnit;
+import org.hisp.dhis.trackedentity.TrackedEntityProgramOwnerStore;
+import org.hisp.dhis.tracker.TrackerIdScheme;
+import org.hisp.dhis.tracker.TrackerImportParams;
+import org.hisp.dhis.tracker.preheat.TrackerPreheat;
+import org.springframework.stereotype.Component;
+
+/**
+ * @author Ameen Mohamed
+ */
+@RequiredArgsConstructor
+@Component
+@SupplierDependsOn( ClassBasedSupplier.class )
+public class ProgramOwnerSupplier extends AbstractPreheatSupplier
+{
+    @NonNull
+    private final TrackedEntityProgramOwnerStore trackedEntityProgramOwnerStore;
+
+    @Override
+    public void preheatAdd( TrackerImportParams params, TrackerPreheat preheat )
+    {
+        final Map<String, TrackedEntityInstance> preheatedTrackedEntities = preheat.getTrackedEntities().getOrDefault(
+            TrackerIdScheme.UID,
+            new HashMap<>() );
+        final Map<String, ProgramInstance> preheatedEnrollments = preheat.getEnrollments().getOrDefault(
+            TrackerIdScheme.UID,
+            new HashMap<>() );
+        Set<Long> teiIds = new HashSet<>();
+        Set<Long> programIds = new HashSet<>();
+        params.getEnrollments().stream().forEach( en -> {
+            TrackedEntityInstance tei = preheatedTrackedEntities.get( en.getTrackedEntity() );
+            if ( tei != null )
+            {
+                teiIds.add( tei.getId() );
+            }
+
+            ProgramInstance pi = preheatedEnrollments.get( en.getEnrollment() );
+            if ( pi != null )
+            {
+                programIds.add( pi.getProgram().getId() );
+            }
+        } );
+
+        params.getEvents().stream().forEach( ev -> {
+            ProgramInstance pi = preheatedEnrollments.get( ev.getEnrollment() );
+            if ( pi != null )
+            {
+                teiIds.add( pi.getEntityInstance().getId() );
+                programIds.add( pi.getProgram().getId() );
+            }
+        } );
+
+        List<TrackedEntityProgramOwnerOrgUnit> tepos = trackedEntityProgramOwnerStore
+            .getTrackedEntityProgramOwnerOrgUnits( teiIds, programIds );
+
```

Comment:
Here we should have something like this:

`tepos = tepos.stream()
                .map(tepo -> new TrackedEntityProgramOwnerOrgUnit(
                tepo.getId(), tepo.getTrackedEntityInstanceId(), tepo.getProgramId(),
                organisationUnitMapper.map(tepo.getOrganisationUnit()) ))
                .collect(Collectors.toList() );`

We need to detach any entity that we are getting from the DB from Hibernate, in order to have everything that we need available at the moment of validation and avoid Lazy Initialisation problems